File: 0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.java
class Solution {
    public int countAtMostK(int nums[], int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        
        int cnt = 0, left = 0, right = 0;
        
        while (right < nums.length) {
            
            map.put(nums[right], map.getOrDefault(nums[right], 0) + 1);
            
            while (map.size() > k) {
                map.put(nums[left], map.getOrDefault(nums[left], 0) - 1);

                if (map.get(nums[left]) <= 0) {
                    map.remove(nums[left]);
                }

                left++;
            }
            
            cnt = cnt + (right - left + 1);
            
            right++;
        }
        
        return cnt;
    }
    
    public int subarraysWithKDistinct(int[] nums, int k) {       
        return countAtMostK(nums, k) - countAtMostK(nums, k - 1);
    }
}

File: 0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.cpp
class Solution {
public:
    int subarray_lte_k(vector<int>& nums, int k) {
        unordered_map<int, int> mp;
        int cnt = 0;
        
        for (int l = 0, r = 0, n = nums.size(); r < n; r++) {
            mp[nums[r]]++;
            
            while (mp.size() > k) {
                mp[nums[l]]--;
                if (mp[nums[l]] == 0) {
                    mp.erase(nums[l]);
                }
                l++;
            }
            
            cnt += (r - l + 1);
        }
        
        return cnt;
    }
    
    int subarraysWithKDistinct(vector<int>& nums, int k) {
        return subarray_lte_k(nums, k) - subarray_lte_k(nums, k - 1);
    }
};